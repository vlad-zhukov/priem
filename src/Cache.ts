const NEXT = '@@next';
const PREV = '@@prev';

export class CacheItem<K, V> {
    key: K;
    value: V;
    [NEXT]: CacheItem<K, V> | undefined;
    [PREV]: CacheItem<K, V> | undefined;
    isValid!: boolean;
    expireTimerId: number | undefined;
    lastUpdateAt: number | undefined;

    constructor(key: K, value: V) {
        this.key = key;
        this.value = value;

        Object.defineProperties(this, {
            [NEXT]: {
                value: undefined,
                writable: true,
            },
            [PREV]: {
                value: undefined,
                writable: true,
            },
            isValid: {
                value: false,
                writable: true,
            },
            expireTimerId: {
                value: undefined,
                writable: true,
            },
            lastUpdateAt: {
                value: undefined,
                writable: true,
            },
        });
    }

    destroy(): void {
        delete this.key;
        delete this.value;
        this[NEXT] = undefined;
        this[PREV] = undefined;
        this.isValid = false;
        clearTimeout(this.expireTimerId);
        this.expireTimerId = undefined;
        this.lastUpdateAt = undefined;
    }
}

export interface SerializableCacheItem<K, V> {
    key: K;
    value: V;
}

const REDUCED = '@@reduced';

interface ReducedType<ValueType> {
    [REDUCED]: true;
    value: ValueType;
}

function reduced<ValueType>(value: ValueType): ReducedType<ValueType> {
    return {[REDUCED]: true, value};
}

function isReduced<ValueType>(value: any): value is ReducedType<ValueType> {
    return value && REDUCED in value;
}

export function reduce<ReturnType, K, V>(
    cache: Cache<K, V>,
    accumulator: ReturnType,
    iteratee: (result: ReturnType, item: CacheItem<K, V>) => ReturnType | ReducedType<ReturnType>,
) {
    let item = cache.head;
    let result: ReturnType | ReducedType<ReturnType> = accumulator;
    while (item) {
        result = iteratee(result, item);
        if (isReduced(result)) {
            return result.value;
        }
        item = item[NEXT];
    }
    return result;
}

export class Cache<K, V> {
    head: CacheItem<K, V> | undefined;
    tail: CacheItem<K, V> | undefined;
    size = 0;

    constructor(items: (CacheItem<K, V> | SerializableCacheItem<K, V>)[]) {
        const now = Date.now();

        for (let i = items.length; i > 0; i--) {
            const item = items[i - 1];
            const cacheItem: CacheItem<K, V> =
                item instanceof CacheItem ? item : new CacheItem<K, V>(item.key, item.value);
            cacheItem.isValid = true;
            cacheItem.lastUpdateAt = now;
            this.prepend(cacheItem);
        }
    }

    prepend(item: CacheItem<K, V>): void {
        item[NEXT] = this.head;
        if (this.head) {
            this.head[PREV] = item;
        }
        this.head = item;
        if (!this.tail) {
            this.tail = item;
        }
        this.size += 1;
    }

    remove(item: CacheItem<K, V>): boolean {
        if (!item.key) {
            return false;
        }

        const next = item[NEXT];
        const prev = item[PREV];

        if (prev) {
            prev[NEXT] = next;
        } else {
            this.head = next;
        }

        if (next) {
            next[PREV] = prev;
        } else {
            this.tail = prev;
        }

        this.size -= 1;

        return true;
    }

    findBy(predicate: (item: CacheItem<K, V>) => boolean): CacheItem<K, V> | undefined {
        return reduce<CacheItem<K, V> | undefined, K, V>(this, undefined, (acc, item) =>
            predicate(item) ? reduced(item) : acc,
        );
    }
}
