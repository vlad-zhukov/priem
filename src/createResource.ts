import * as React from 'react';
import {TypeName} from '@sindresorhus/is';
import {Resource, ResourceOptions, Subscriber, Status, MemoizedKey} from './Resource';
import {assertType, isBrowser, shallowEqual, useForceUpdate, useLazyRef} from './utils';

const DEFAULT_DEBOUNCE_MS = 150;

export interface Options {
    maxAge?: number;
    refreshOnMount?: boolean;
}

export interface ResultMeta {
    pending: boolean;
    fulfilled: boolean;
    rejected: boolean;
    reason: Error | undefined;
    invalidate: () => void;
}

export type Result<DataType> = [DataType | undefined, ResultMeta];

interface Refs<Args, DataType> extends Subscriber<Args> {
    args?: Args;
    lastTimeCalled: number;
    prevResult?: Result<DataType>;
}

export function createResource<DataType, Args extends MemoizedKey>(
    fn: (args: Args) => Promise<DataType>,
    resourceOptions: ResourceOptions = {},
) {
    const resource = new Resource<DataType, Args>(fn, resourceOptions);

    return function useResource(args: Args | undefined, options: Options = {}): Result<DataType> {
        assertType(args, [TypeName.Object, TypeName.undefined], '`args`');

        const forceUpdate = useForceUpdate();

        const {current: refs} = React.useRef<Refs<Args, DataType>>({
            /* istanbul ignore next */
            onChange() {
                // A callback for Resource#onCacheChange
                return false;
            },
            lastTimeCalled: 0,
        });

        refs.args = args;
        refs.onChange = function onChange(prevArgs, shouldCommit) {
            if (refs.args && prevArgs && shallowEqual(refs.args, prevArgs)) {
                if (shouldCommit) {
                    forceUpdate();
                }
                return true;
            }
            return false;
        };

        useLazyRef(() => {
            if (!!options.refreshOnMount && refs.args) {
                resource.delete(refs.args);
            }
            resource.subscribe(refs);
        });

        React.useEffect(() => {
            return () => {
                resource.unsubscribe(refs);
            };
        }, []); // eslint-disable-line react-hooks/exhaustive-deps

        const {lastTimeCalled, prevResult} = refs;
        const now = Date.now();
        refs.lastTimeCalled = now;

        /**
         * Should this call get debounced and rescheduled,
         * and return the previous value to reduce the amount of requests?
         *
         * We should debounce when all conditions are met:
         * 1. Arguments are provided.
         * 2. Previous result is valid.
         * 3. Less than 150ms lapsed since the last call.
         * 4. The item is not in the cache.
         */
        const shouldDebounce =
            isBrowser &&
            args !== undefined &&
            !!prevResult &&
            now - lastTimeCalled < DEFAULT_DEBOUNCE_MS &&
            !resource.has(args);

        // TODO: rework debounce
        React.useEffect(() => {
            let handler: number | undefined;
            if (shouldDebounce) {
                handler = window.setTimeout(forceUpdate, DEFAULT_DEBOUNCE_MS);
            }
            return () => window.clearTimeout(handler);
        });

        if (shouldDebounce) {
            return prevResult as Result<DataType>;
        }

        const meta: ResultMeta = {
            pending: false,
            fulfilled: false,
            rejected: false,
            reason: undefined,
            invalidate() {
                if (refs.args) {
                    resource.invalidate(refs.args, false);
                    resource.read(refs.args, {maxAge: options.maxAge});
                }
            },
        };

        if (args === undefined) {
            return [undefined, meta];
        }

        const ret = resource.read(args, {maxAge: options.maxAge});

        if ((!ret || ret.status === Status.PENDING) && !!prevResult) {
            return prevResult;
        }

        let data = prevResult ? prevResult[0] : undefined;

        if (ret) {
            if (ret.data) {
                data = ret.data;
            }
            meta.pending = ret.status === Status.PENDING;
            meta.fulfilled = ret.status === Status.FULFILLED;
            meta.rejected = ret.status === Status.REJECTED;
            meta.reason = ret.reason;
        }

        const result: Result<DataType> = [data, meta];
        refs.prevResult = result;

        return result;
    };
}
