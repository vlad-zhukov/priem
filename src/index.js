export {default as Priem} from './Priem';
export {Container, populateStore, flushStore} from './Container';
export {default as withPriem} from './withPriem';
export {default as getDataFromTree} from './getDataFromTree';
