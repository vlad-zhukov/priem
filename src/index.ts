export {createResource, GetArgs, Options, Result, ResultMeta, ResultPages, ResultPagesMeta} from './createResource';
export {
    hydrateStore,
    flushStore,
    getRunningPromises,
    ResourceOptions,
    MemoizedKey,
    MemoizedValue,
    MemoizedSerializableCacheItem,
    Status,
} from './Resource';
export {SerializableCacheItem} from './Cache';
