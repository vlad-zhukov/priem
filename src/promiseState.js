import {type} from './helpers';

/**
 * Constructor for creating a new promiseState.
 * @param [options] {Object}
 * @param [options.pending] {Boolean}
 * @param [options.refreshing] {Boolean}
 * @param [options.fulfilled] {Boolean}
 * @param [options.rejected] {Boolean}
 * @param [options.value] {*}
 * @param [options.reason] {String|Null}
 * @returns {Object}
 */
function create({
    pending = false, // eslint-disable-line no-shadow
    refreshing = false, // eslint-disable-line no-shadow
    fulfilled = false, // eslint-disable-line no-shadow
    rejected = false, // eslint-disable-line no-shadow
    value = null,
    reason = null,
}) {
    return {
        pending,
        refreshing,
        fulfilled,
        rejected,
        value,
        reason,
    };
}

/**
 * Checks if a passed value is a promiseState-like object.
 * @param maybePromiseState {*}
 * @returns {Boolean}
 */
export function isPromiseState(maybePromiseState) {
    if (type(maybePromiseState) !== 'object') {
        return false;
    }

    const propertyNames = ['pending', 'refreshing', 'fulfilled', 'rejected'];
    return propertyNames.every(propertyName => type(maybePromiseState[propertyName]) === 'boolean');
}

/**
 * Checks if a passed promiseState is in a loading state.
 * @param promiseState {*}
 * @returns {Boolean}
 */
export function isLoading(promiseState) {
    if (isPromiseState(promiseState) === false) {
        throw new TypeError('Priem: Expected a promiseState.');
    }
    return promiseState.pending || promiseState.refreshing;
}

/**
 * Creates a new promiseState that is pending.
 * @returns {Object}
 */
export function pending() {
    return create({pending: true});
}

/**
 * Creates a promiseState that is refreshing.
 * Can be called without a previous promiseState and will be both pending and refreshing.
 * @param [previous] {Object}
 * @returns {Object}
 */
export function refreshing(previous) {
    const previousPromiseState = isPromiseState(previous) ? previous : pending();
    return create({...previousPromiseState, refreshing: true});
}

/**
 * Creates a promiseState that is resolved with the given value.
 * If the given value is already a promiseState, its `value` property will be used instead.
 * @param [valueOrPromiseState] {*}
 * @returns {Object}
 */
export function fulfilled(valueOrPromiseState) {
    const value = isPromiseState(valueOrPromiseState) ? valueOrPromiseState.value : valueOrPromiseState;
    return create({fulfilled: true, value});
}

/**
 * Creates a promiseState that is rejected with the given reason.
 * @param reason {String}
 * @returns {Object}
 */
export function rejected(reason) {
    return create({rejected: true, reason});
}
