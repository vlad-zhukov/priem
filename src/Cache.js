import moize from 'moize';
import {elementsEqual} from 'react-shallow-equal';
import * as promiseState from './promiseState';
import {type, assertType, isBrowser} from './helpers';

function arrayElementsEqual(a, b) {
    if (!Array.isArray(a) || !Array.isArray(b) || a.length !== b.length) {
        return false;
    }
    for (let i = 0; i < a.length; i++) {
        if (a[i] !== b[i]) {
            return false;
        }
    }

    return true;
}

export const defaultAsyncValueOptions = {autoRefresh: true, maxSize: Infinity};

export default class Cache {
    constructor() {
        this.memoized = null;
        this.awaiting = [];
        this.rejected = false;
        this.prevArgs = null;
    }

    add(value, onExpire) {
        assertType(value, ['object'], "'getAsyncValue() -> value'");
        assertType(value.promise, ['function'], "'getAsyncValue() -> value.promise'");

        this.memoized = moize(value.promise, {
            isPromise: true,
            maxAge: value.maxAge,
            maxArgs: value.maxArgs,
            maxSize: value.maxSize,
            onExpire,
        });
    }

    isMemoized(args) {
        if (!this.memoized) {
            return false;
        }

        return this.memoized.has(args);
    }

    getAwaitingIndex(args) {
        for (let i = 0, l = this.awaiting.length; i < l; i++) {
            if (elementsEqual(this.awaiting[i], args)) {
                return i;
            }
        }

        return -1;
    }

    isAwaiting(args) {
        return this.getAwaitingIndex(args) !== -1;
    }

    removeAwaiting(args) {
        const index = this.getAwaitingIndex(args);
        if (index !== -1) {
            this.awaiting.splice(index, 1);
        }
    }

    run({asyncValue, isForced, update, onExpire}) {
        const value = {...defaultAsyncValueOptions, ...asyncValue};

        // Stop auto-refreshing
        if (!value.autoRefresh && !isForced) {
            return;
        }

        const args = type(value.args) === 'array' ? value.args : [];

        if (!this.memoized) {
            this.add(value, onExpire);
        }

        // Do not recall rejected (uncached) promises unless forced
        if (this.rejected && !isForced) {
            return;
        }

        if (this.isAwaiting(args)) {
            return;
        }

        let shouldUpdateState = true;

        // eslint-disable-next-line consistent-return
        update((s, m) => {
            const isFulfilled = !!s.fulfilled;
            const isSsr = !!(isFulfilled && m.ssr);

            if (this.isMemoized(args)) {
                // Do not recall memoized promises unless forced
                if (isFulfilled && !isForced && arrayElementsEqual(args, this.prevArgs)) {
                    shouldUpdateState = false;
                    return null;
                }
            }
            else if (isSsr) {
                shouldUpdateState = false;
                this.memoized.add(args, Promise.resolve(s.value));
                this.prevArgs = args;
                // eslint-disable-next-line consistent-return
                return {
                    meta: {ssr: false},
                };
            }

            this.awaiting.unshift(args);
            this.rejected = false;

            if (isFulfilled) {
                return {
                    state: promiseState.refreshing(s),
                };
            }

            return {
                state: promiseState.pending(),
            };
        });

        if (!shouldUpdateState) {
            return;
        }

        if (isForced) {
            this.memoized.remove(args);
        }

        // eslint-disable-next-line consistent-return
        return this.memoized(...args)
            .then((result) => {
                this.removeAwaiting(args);
                this.prevArgs = args;
                update({
                    state: promiseState.fulfilled(result),
                    meta: {ssr: !isBrowser},
                });
            })
            .catch((e) => {
                this.removeAwaiting(args);
                this.rejected = true;
                this.prevArgs = null;
                update({
                    state: promiseState.rejected(e.message),
                    meta: {ssr: !isBrowser},
                });
            });
    }
}
