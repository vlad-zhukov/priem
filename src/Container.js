import memoize, {FULFILLED, REJECTED} from './memoize';
import * as promiseState from './promiseState';
import {assertType} from './helpers';

export default class Container {
    constructor(options) {
        assertType(options, ['object'], "AsyncContainer argument 'options'");
        assertType(options.promise, ['function'], "'promise'");
        assertType(options.mapPropsToArgs, ['function', 'undefined'], "'mapPropsToArgs'");

        this._mapPropsToArgs = options.mapPropsToArgs || (() => []);
        this._listeners = [];
        this._recentCallCount = 0;
        this._lastCallTime = 0;
        this._prevProps = {};

        this._notify = this._notify.bind(this);

        this._memoized = memoize(options.promise, {
            maxSize: options.maxSize,
            maxAge: options.maxAge,
            onCacheChange: this._notify,
        });

        this._debouncedGet = function(props = this._prevProps, isForced = false) {
            this._prevProps = props;

            const args = this._mapPropsToArgs(props);
            assertType(args, ['array', 'null'], "The result of 'mapPropsToArgs(props)'");

            return this._memoized.apply({isForced}, args);
        };
    }

    _get(props, isForced) {
        const now = Date.now();
        if (now - this._lastCallTime < 200) {
            if (this._recentCallCount > 100) {
                throw new Error(
                    "Priem: the amount of updates of 'Container' exceeded the safe threshold, which means " +
                        "it has stuck in an infinite rerendering loop. This happens when 'mapPropsToArgs' " +
                        'returns different results on consecutive calls. For example, this might be caused by ' +
                        'a race condition between 2 or more Priem components. Please, fix.'
                );
            }
            this._recentCallCount += 1;
        } else {
            this._recentCallCount = 1;
        }
        this._lastCallTime = now;

        //

        const res = this._debouncedGet(props, isForced);

        console.log(res);

        switch (res.status) {
            case FULFILLED:
                return promiseState.fulfilled(res.value);
            case REJECTED:
                return promiseState.rejected(res.reason);
            default:
                return promiseState.pending();
        }
    }

    _notify() {
        this._listeners.forEach(fn => fn());
    }

    _subscribe(fn) {
        this._listeners.push(fn);
    }

    _unsubscribe(fn) {
        this._listeners = this._listeners.filter(f => f !== fn);
    }

    refresh() {
        const res = this._debouncedGet(this._prevProps, true);
        return res.promise;
    }
}
