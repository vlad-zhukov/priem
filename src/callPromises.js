import moize from 'moize';
import {elementsEqual} from 'react-shallow-equal';
import * as promiseState from './promiseState';
import {type} from './helpers';

const memoized = {};

function memoizeAsyncValue(key, value, onExpire) {
    const typeOfValue = type(value);
    if (typeOfValue !== 'object') {
        throw new TypeError(`Priem: argument 'values' must return an object of objects, but got: '${typeOfValue}'.`);
    }

    const typeOfValuePromise = type(value.promise);
    if (typeOfValuePromise !== 'function') {
        throw new TypeError(
            "Priem: argument 'values' must return an object of objects with a property 'promise' " +
                `each, but got: '${typeOfValuePromise}'.`
        );
    }

    memoized[key] = moize(value.promise, {
        isPromise: true,
        maxAge: value.maxAge,
        maxArgs: value.maxArgs,
        maxSize: value.maxSize,
        onExpire,
    });
}

export function extractAsyncValues(props) {
    if (type(props.asyncValues) !== 'function') {
        return {};
    }

    const values = props.asyncValues(props);

    const typeOfValues = type(values);
    if (typeOfValues !== 'object') {
        throw new TypeError(
            `Priem: property 'asyncValues' must be a function that returns an object, but got: '${typeOfValues}'.`
        );
    }

    return values;
}

function callPromise({props, key, asyncValue, prevArgs, onExpire, isMounting, isForced}) {
    // Do not recall rejected (uncached) promises unless forced
    if (
        isForced === false &&
        props.priem !== undefined &&
        props.priem[key] !== undefined &&
        props.priem[key].rejected === true
    ) {
        return;
    }

    const args = type(asyncValue.args) === 'array' ? asyncValue.args : [];

    if (memoized[key] !== undefined) {
        if (memoized[key].has(args) === true) {
            if (isMounting === false && elementsEqual(args, prevArgs)) {
                return;
            }
        }
        else {
            // Only set to refreshing when the result is not cached
            props.setPriem(s => ({
                [key]: promiseState.refreshing(s[key]),
            }));
        }
    }
    else {
        memoizeAsyncValue(key, asyncValue, onExpire);

        if (isMounting === false) {
            props.setPriem({
                [key]: promiseState.pending(),
            });
        }
    }

    memoized[key](...args)
        .then((result) => {
            props.setPriem({
                [key]: promiseState.fulfilled(result),
            });
        })
        .catch((e) => {
            props.setPriem({
                [key]: promiseState.rejected(e.message),
            });
        });
}

export function callPromises({props, prevProps, onExpire, isMounting = false, isForced = false}) {
    // Stop auto-refreshing
    if (props.autoRefresh === false && isForced === false) {
        return;
    }

    const asyncValues = extractAsyncValues(props);
    const prevAsyncValues = prevProps && prevProps.priem ? extractAsyncValues(prevProps) : null;
    const asyncKeys = Object.keys(asyncValues);
    for (let i = 0, l = asyncKeys.length; i < l; i++) {
        const key = asyncKeys[i];
        const asyncValue = asyncValues[key];

        const prevArgs =
            prevAsyncValues && type(prevAsyncValues[key].args) === 'array' ? prevAsyncValues[key].args : [];

        callPromise({props, key, asyncValue, prevArgs, onExpire, isMounting, isForced});
    }
}
