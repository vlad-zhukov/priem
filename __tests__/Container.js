import delay from 'delay';
import {AsyncContainer} from '../src/Container';
import * as promiseState from '../src/promiseState';

function setup({props = {}, getAsyncValue, ...options}) {
    const container = new AsyncContainer(getAsyncValue);

    const updateSpy = jest.spyOn(container, 'update');
    const runAsyncSpy = jest.spyOn(container, 'runAsync');
    const subscribeSpy = jest.fn(() => {});
    container.subscribe(subscribeSpy);

    function runAsync(moreOptions) {
        return container.runAsync({props, ...options, ...moreOptions});
    }

    return {
        container,
        updateSpy,
        runAsyncSpy,
        subscribeSpy,
        runAsync,
    };
}

it('should not run promises if both `autoRefresh` and `isForced` are false', async () => {
    const getAsyncValue = () => ({
        args: ['foo'],
        promise: value => delay(200, {value}),
        autoRefresh: false,
    });

    const {container, updateSpy, runAsyncSpy, runAsync} = setup({getAsyncValue, isForced: false});

    runAsync();
    expect(container.cache.awaiting).toMatchObject([]);

    await delay(250);
    expect(container.cache.awaiting).toMatchObject([]);

    expect(updateSpy).toHaveBeenCalledTimes(0);
    expect(runAsyncSpy).toHaveBeenCalledTimes(1);
    expect([container.state, container.meta]).toMatchSnapshot();
});

it('should run promises if `autoRefresh` is false but `isForced` is true', async () => {
    const getAsyncValue = () => ({
        args: ['foo'],
        promise: value => delay(200, {value}),
        autoRefresh: false,
    });

    const {container, updateSpy, runAsyncSpy, runAsync} = setup({getAsyncValue, isForced: true});

    runAsync();
    expect(container.cache.awaiting).toMatchObject([['foo']]);

    await delay(250);
    expect(container.cache.awaiting).toMatchObject([]);

    expect(updateSpy).toHaveBeenCalledTimes(2);
    expect(runAsyncSpy).toHaveBeenCalledTimes(1);
    expect([container.state, container.meta]).toMatchSnapshot();
});

it('should not run promises when awaiting', async () => {
    const getAsyncValue = () => ({
        args: ['foo'],
        promise: value => delay(200, {value}),
    });

    const {container, updateSpy, runAsyncSpy, runAsync} = setup({getAsyncValue, isForced: false});

    runAsync();
    expect(container.cache.awaiting).toMatchObject([['foo']]);

    runAsync();
    expect(container.cache.awaiting).toMatchObject([['foo']]);

    runAsync();
    expect(container.cache.awaiting).toMatchObject([['foo']]);

    await delay(250);
    expect(container.cache.awaiting).toMatchObject([]);

    expect(updateSpy).toHaveBeenCalledTimes(2);
    expect(runAsyncSpy).toHaveBeenCalledTimes(3);
    expect([container.state, container.meta]).toMatchSnapshot();
});

it('should refresh a promise if forced', async () => {
    let counter = 0;
    const getAsyncValue = () => ({
        args: ['foo'],
        promise(value) {
            counter += 1;
            return delay(200, value + counter);
        },
    });

    const {container, updateSpy, runAsync} = setup({getAsyncValue, isForced: false});

    await runAsync();
    expect([container.state, container.meta]).toMatchSnapshot();

    runAsync({isForced: true});
    expect([container.state, container.meta]).toMatchSnapshot();
    await delay(250);

    expect([container.state, container.meta]).toMatchSnapshot();

    expect(updateSpy).toHaveBeenCalledTimes(4);
});

it('should not rerun promises if previous promise was rejected', async () => {
    let called = false;
    const getAsyncValue = () => ({
        args: ['foo'],
        promise(value) {
            if (!called) {
                called = true;
                return delay.reject(200, new Error('foo'));
            }
            return delay(200, {value});
        },
    });

    const {container, updateSpy, runAsyncSpy, runAsync} = setup({getAsyncValue, isForced: false});

    runAsync();
    expect(container.cache.awaiting).toMatchObject([['foo']]);

    await delay(250);
    expect(container.cache.awaiting).toMatchObject([]);

    runAsync();
    expect(container.cache.awaiting).toMatchObject([]);

    expect(updateSpy).toHaveBeenCalledTimes(2);
    expect(runAsyncSpy).toHaveBeenCalledTimes(2);
    expect([container.state, container.meta]).toMatchSnapshot();
});

it('should rerun promises if previous promise was rejected but `isForced` is true', async () => {
    let called = false;
    const getAsyncValue = () => ({
        args: ['foo'],
        promise(value) {
            if (!called) {
                called = true;
                return delay.reject(200, new Error('foo'));
            }
            return delay(200, {value});
        },
    });

    const {container, updateSpy, runAsyncSpy, runAsync} = setup({getAsyncValue, isForced: true});

    runAsync();
    expect(container.cache.awaiting).toMatchObject([['foo']]);

    await delay(250);
    expect(container.cache.awaiting).toMatchObject([]);

    runAsync();
    expect(container.cache.awaiting).toMatchObject([['foo']]);

    await delay(250);
    expect(container.cache.awaiting).toMatchObject([]);

    expect(updateSpy).toHaveBeenCalledTimes(4);
    expect(runAsyncSpy).toHaveBeenCalledTimes(2);
    expect([container.state, container.meta]).toMatchSnapshot();
});

it('should not try to rerun fulfilled promises if `value` is null', async () => {
    const getAsyncValue = () => ({
        args: [null],
        promise: value => delay(100, value),
    });

    const {container, updateSpy, runAsyncSpy, runAsync} = setup({getAsyncValue, isForced: false});

    runAsync();
    expect(container.cache.awaiting).toMatchObject([[null]]);

    await delay(150);
    expect(container.cache.awaiting).toMatchObject([]);

    runAsync();
    expect(container.cache.awaiting).toMatchObject([]);

    expect(updateSpy).toHaveBeenCalledTimes(3);
    expect(runAsyncSpy).toHaveBeenCalledTimes(2);
    expect([container.state, container.meta]).toMatchSnapshot();
});

it('should expire if maxAge is set', async () => {
    const getAsyncValue = () => ({
        args: ['foo'],
        promise: value => delay(200, {value}),
        maxAge: 300,
    });

    const {container, updateSpy, runAsyncSpy, runAsync} = setup({getAsyncValue, isForced: false});

    runAsync();
    expect(container.cache.awaiting).toMatchObject([['foo']]);

    await delay(250);
    expect([container.state, container.meta]).toMatchSnapshot(); // fulfilled

    await delay(100);
    expect([container.state, container.meta]).toMatchSnapshot(); // refreshing

    await delay(650);
    expect([container.state, container.meta]).toMatchSnapshot(); // refreshing
    expect(container.cache.awaiting).toMatchObject([['foo']]);

    expect(updateSpy).toHaveBeenCalledTimes(7);
    expect(runAsyncSpy).toHaveBeenCalledTimes(4);
});

it('should rehydrate ssr data', async () => {
    const promiseFn = jest.fn(value => delay(200, {value}));

    const getAsyncValue = () => ({
        args: ['foo'],
        promise: promiseFn,
    });

    const {container, updateSpy, runAsyncSpy, runAsync} = setup({getAsyncValue, isForced: false});

    container.update({
        data: promiseState.fulfilled('foo'),
        meta: {ssr: true},
    });

    await runAsync();

    expect([container.state, container.meta]).toMatchSnapshot();
    expect(await container.cache.memoized.get(['foo'])).toBe('foo');

    expect(promiseFn).toHaveBeenCalledTimes(0);
    expect(updateSpy).toHaveBeenCalledTimes(3);
    expect(runAsyncSpy).toHaveBeenCalledTimes(1);
});

it('should add values to cache when `args` change', async () => {
    let id = 0;
    const getAsyncValue = () => ({
        args: [`foo${id}`, `bar${id}`],
        promise(foo, bar) {
            id += 1;
            return delay(200, {foo, bar});
        },
    });

    const {container, updateSpy, runAsync} = setup({getAsyncValue, isForced: false});

    await runAsync();
    await runAsync();

    expect(await container.cache.memoized.keys()).toEqual([['foo1', 'bar1'], ['foo0', 'bar0']]);

    expect(updateSpy).toHaveBeenCalledTimes(4);
});

it('should return cached values when `args` change', async () => {
    let check = false;
    const getAsyncValue = () => ({
        args: [`foo-${check}`],
        promise(value) {
            check = !check;
            return delay(200, {value});
        },
    });

    const {container, runAsync} = setup({getAsyncValue, isForced: false});

    await runAsync();
    expect([container.state, container.meta]).toMatchSnapshot(); // foo-false

    await runAsync();
    expect([container.state, container.meta]).toMatchSnapshot(); // foo-true

    await runAsync();
    expect([container.state, container.meta]).toMatchSnapshot(); // foo-false
});
