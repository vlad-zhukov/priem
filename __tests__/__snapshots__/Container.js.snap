// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should expire if maxAge is set 1`] = `
Array [
  Object {
    "fulfilled": true,
    "pending": false,
    "reason": null,
    "refreshing": false,
    "rejected": false,
    "value": Object {
      "value": "foo",
    },
  },
  Object {
    "args": Array [
      "foo",
    ],
    "ssr": false,
  },
]
`;

exports[`should expire if maxAge is set 2`] = `
Array [
  Object {
    "fulfilled": true,
    "pending": false,
    "reason": null,
    "refreshing": true,
    "rejected": false,
    "value": Object {
      "value": "foo",
    },
  },
  Object {
    "args": Array [
      "foo",
    ],
    "ssr": false,
  },
]
`;

exports[`should expire if maxAge is set 3`] = `
Array [
  Object {
    "fulfilled": true,
    "pending": false,
    "reason": null,
    "refreshing": true,
    "rejected": false,
    "value": Object {
      "value": "foo",
    },
  },
  Object {
    "args": Array [
      "foo",
    ],
    "ssr": false,
  },
]
`;

exports[`should not rerun promises if previous promise was rejected 1`] = `
Array [
  Object {
    "fulfilled": false,
    "pending": false,
    "reason": "foo",
    "refreshing": false,
    "rejected": true,
    "value": null,
  },
  Object {
    "args": null,
    "ssr": false,
  },
]
`;

exports[`should not run promises if both \`autoRefresh\` and \`isForced\` are false 1`] = `
Array [
  Object {
    "fulfilled": false,
    "pending": true,
    "reason": null,
    "refreshing": false,
    "rejected": false,
    "value": null,
  },
  Object {
    "ssr": false,
  },
]
`;

exports[`should not run promises when awaiting 1`] = `
Array [
  Object {
    "fulfilled": true,
    "pending": false,
    "reason": null,
    "refreshing": false,
    "rejected": false,
    "value": Object {
      "value": "foo",
    },
  },
  Object {
    "args": Array [
      "foo",
    ],
    "ssr": false,
  },
]
`;

exports[`should not try to rerun fulfilled promises if \`value\` is null 1`] = `
Array [
  Object {
    "fulfilled": true,
    "pending": false,
    "reason": null,
    "refreshing": false,
    "rejected": false,
    "value": null,
  },
  Object {
    "args": Array [
      null,
    ],
    "ssr": false,
  },
]
`;

exports[`should refresh a promise if forced 1`] = `
Array [
  Object {
    "fulfilled": true,
    "pending": false,
    "reason": null,
    "refreshing": false,
    "rejected": false,
    "value": "foo1",
  },
  Object {
    "args": Array [
      "foo",
    ],
    "ssr": false,
  },
]
`;

exports[`should refresh a promise if forced 2`] = `
Array [
  Object {
    "fulfilled": true,
    "pending": false,
    "reason": null,
    "refreshing": true,
    "rejected": false,
    "value": "foo1",
  },
  Object {
    "args": Array [
      "foo",
    ],
    "ssr": false,
  },
]
`;

exports[`should refresh a promise if forced 3`] = `
Array [
  Object {
    "fulfilled": true,
    "pending": false,
    "reason": null,
    "refreshing": false,
    "rejected": false,
    "value": "foo2",
  },
  Object {
    "args": Array [
      "foo",
    ],
    "ssr": false,
  },
]
`;

exports[`should rehydrate ssr data 1`] = `
Array [
  Object {
    "fulfilled": true,
    "pending": false,
    "reason": null,
    "refreshing": false,
    "rejected": false,
    "value": "foo",
  },
  Object {
    "args": Array [
      "foo",
    ],
    "ssr": false,
  },
]
`;

exports[`should rerun promises if previous promise was rejected but \`isForced\` is true 1`] = `
Array [
  Object {
    "fulfilled": true,
    "pending": false,
    "reason": null,
    "refreshing": false,
    "rejected": false,
    "value": Object {
      "value": "foo",
    },
  },
  Object {
    "args": Array [
      "foo",
    ],
    "ssr": false,
  },
]
`;

exports[`should return cached values when \`args\` change 1`] = `
Array [
  Object {
    "fulfilled": true,
    "pending": false,
    "reason": null,
    "refreshing": false,
    "rejected": false,
    "value": Object {
      "value": "foo-false",
    },
  },
  Object {
    "args": Array [
      "foo-false",
    ],
    "ssr": false,
  },
]
`;

exports[`should return cached values when \`args\` change 2`] = `
Array [
  Object {
    "fulfilled": true,
    "pending": false,
    "reason": null,
    "refreshing": false,
    "rejected": false,
    "value": Object {
      "value": "foo-true",
    },
  },
  Object {
    "args": Array [
      "foo-true",
    ],
    "ssr": false,
  },
]
`;

exports[`should return cached values when \`args\` change 3`] = `
Array [
  Object {
    "fulfilled": true,
    "pending": false,
    "reason": null,
    "refreshing": false,
    "rejected": false,
    "value": Object {
      "value": "foo-false",
    },
  },
  Object {
    "args": Array [
      "foo-false",
    ],
    "ssr": false,
  },
]
`;

exports[`should run promises if \`autoRefresh\` is false but \`isForced\` is true 1`] = `
Array [
  Object {
    "fulfilled": true,
    "pending": false,
    "reason": null,
    "refreshing": false,
    "rejected": false,
    "value": Object {
      "value": "foo",
    },
  },
  Object {
    "args": Array [
      "foo",
    ],
    "ssr": false,
  },
]
`;
